x/tokenfactory/keeper/before_send_test.go

```diff
@@ -115,3 +241,99 @@ func (suite *KeeperTestSuite) TestNonWhitelistedHooksNotCalled() {
 	// Whitelisted contract is not called
 	suite.Require().Equal("\"0\"", string(queryResp))
 }
+
+// TestInfiniteTrackBeforeSend tests gas metering with infinite loop contract
+// to properly test if we are gas metering trackBeforeSend properly.
+func (suite *KeeperTestSuite) TestInfiniteTrackBeforeSend() {
+	for _, tc := range []struct {
+		name            string
+		wasmFile        string
+		tokenToSend     sdk.Coins
+		useFactoryDenom bool
+		blockBeforeSend bool
+		expectedError   bool
+	}{
+		{
+			name:            "sending tokenfactory denom from module to module with infinite contract should panic",
+			wasmFile:        "./testdata/infinite_track_beforesend.wasm",
+			useFactoryDenom: true,
+		},
+		{
+			name:            "sending tokenfactory denom from account to account with infinite contract should panic",
+			wasmFile:        "./testdata/infinite_track_beforesend.wasm",
+			useFactoryDenom: true,
+			blockBeforeSend: true,
+		},
+		{
+			name:            "sending non-tokenfactory denom from module to module with infinite contract should not panic",
+			wasmFile:        "./testdata/infinite_track_beforesend.wasm",
+			tokenToSend:     sdk.NewCoins(sdk.NewInt64Coin("foo", 1000000)),
+			useFactoryDenom: false,
+		},
+		{
+			name:            "Try using no 100",
+			wasmFile:        "./testdata/no100.wasm",
```

Comment:
Not sure i understand, what the contract does, could you give a brief description?